# SWS3009 Trial Lecture - Neural Networks

Please fill in your student number and name below. Each student is to make one submission.

| Student Number | Name                 |
|:--------------:|:--------------------:|
| A1234567A      | Jack O'Lantern       |

**Submission:**

Please upload to the respective LumiNUS folders for your group by Friday 11 September 2359 hours for the Tuesday 2 pm lab group, and Monday 2359 hours for the Friday 12 pm lab group.


## 1. Introduction

The objectives of this lab are:

    1. To familiarize you with how to create dense neural networks using Keras.
    2. To familiarize you with how to encode input and output vectors for neural networks.
    3. To give you some insight into how hyperparameters like learning rate and momentum affect training.
    
To save time we will train each experiment only for 50 epochs. This will lead to less than optimal results but is enough for you to make observations.

**HINT: YOU CAN HIT SHIFT-ENTER TO RUN EACH CELL. NOTE THAT IF A CELL IS DEPENDENT ON A PREVIOUS CELL, YOU WILL NEED TO RUN THE PREVIOUS CELL(S) FIRST **


## 2. The Irises Dataset

We will now work again on the Irises Dataset, which we used in Lab 1, for classifying iris flowers into one of three possible types. As before we will consider four factors:

    1. Sepal length in cm
    2. Sepal width in cm
    3. Petal length in cm
    4. Petal width in cm

In this dataset there are 150 sample points. The code below loads the dataset and prints the first 10 rows so we have an idea of what it looks like.

In [16]:
from sklearn.datasets import load_iris
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

iris = load_iris()

print("First 10 rows of data:")
print(iris.data[:10])

First 10 rows of data:
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]]


### 2.2 Scaling the Data

We make use of the MinMaxScaler to scale the inputs to between 0 and 1.  The code below does this and prints the first 10 rows again, to show us the difference.

In the next section we will investigate what happens if we use unscaled data.

In [17]:
scaler = MinMaxScaler()
scaler.fit(iris.data)
X = scaler.transform(iris.data)

print("First 10 rows of SCALED data.")
print(X[:10])

First 10 rows of SCALED data.
[[0.22222222 0.625      0.06779661 0.04166667]
 [0.16666667 0.41666667 0.06779661 0.04166667]
 [0.11111111 0.5        0.05084746 0.04166667]
 [0.08333333 0.45833333 0.08474576 0.04166667]
 [0.19444444 0.66666667 0.06779661 0.04166667]
 [0.30555556 0.79166667 0.11864407 0.125     ]
 [0.08333333 0.58333333 0.06779661 0.08333333]
 [0.19444444 0.58333333 0.08474576 0.04166667]
 [0.02777778 0.375      0.06779661 0.04166667]
 [0.16666667 0.45833333 0.08474576 0.        ]]


### 2.3 Encoding the Targets

In Lab 1 we saw that the target values (type of iris flower) is a vector from 0 to 2. We can see the 150 labels below:


In [18]:
print(iris.target)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


We can use this to train the neural network, but we will use "one-hot" encoding, where we have a vector of _n_ integers consisting of 0's and 1's.  The table below shows how one-hot encoding works:

|   Value    |    One-Hot Encoding    |
|:----------:|:----------------------:|
| 0 | \[1 0 0\] |
| 1 | \[0 1 0\] |
| 2 | \[0 0 1\] |

Keras provides the to_categorical function to create one-hot vectors:



In [19]:
from tensorflow.keras.utils import to_categorical

Y = to_categorical(y = iris.target, num_classes = 3)
print(Y)

[[1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0.

Now let's split the data into training and testing data:



In [20]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, 
                                                    random_state = 1)


### 2.4 Building our Neural Network

Let's now begin building a simple neural network with a single hidden layer, using the Stochastic Gradient Descent (SGD) optimizer, ReLu transfer functions for the hidden layer and softmax for the output layer.

The code to do this is shown below:

In [21]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import SGD
from tensorflow.keras.utils import to_categorical

# Create the neural network
nn = Sequential()
nn.add(Dense(100, input_shape = (4, ), activation = 'relu'))
nn.add(Dense(3, activation = 'softmax'))

# Create our optimizer
sgd = SGD(learning_rate = 0.1, momentum = 1.0)

# 'Compile' the network to associate it with a loss function,
# an optimizer, and what metrics we want to track
nn.compile(loss='categorical_crossentropy', optimizer=sgd, 
          metrics = 'accuracy')

### 2.5 Training the Neural Network

As is usually the case, we can call the "fit" method to train the neural network for 50 epochs. We will shuffle the training data between epochs, and provide validation data.

In [22]:
nn.fit(X_train, Y_train, shuffle = True, epochs = 50, 
      validation_data = (X_test, Y_test))

Epoch 1/50


2022-05-08 16:19:34.885257: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:113] Plugin optimizer for device_type GPU is enabled.


4/4 [==============================] - 2s 39ms/step - loss: 1.0594 - accuracy: 0.5750 - val_loss: 1.0293 - val_accuracy: 0.5000
Epoch 2/50
4/4 [==============================] - 0s 10ms/step - loss: 0.8947 - accuracy: 0.6667 - val_loss: 0.8649 - val_accuracy: 0.5667
Epoch 3/50
4/4 [==============================] - 0s 11ms/step - loss: 0.7109 - accuracy: 0.7083 - val_loss: 0.5666 - val_accuracy: 0.7667
Epoch 4/50
4/4 [==============================] - 0s 10ms/step - loss: 0.4955 - accuracy: 0.8167 - val_loss: 0.4499 - val_accuracy: 0.6667
Epoch 5/50
4/4 [==============================] - 0s 10ms/step - loss: 0.3548 - accuracy: 0.8583 - val_loss: 0.2612 - val_accuracy: 0.9667
Epoch 6/50
1/4 [======>.......................] - ETA: 0s - loss: 0.2923 - accuracy: 0.9688

2022-05-08 16:19:36.615586: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:113] Plugin optimizer for device_type GPU is enabled.


4/4 [==============================] - 0s 11ms/step - loss: 0.2614 - accuracy: 0.9333 - val_loss: 0.2212 - val_accuracy: 0.9667
Epoch 7/50
4/4 [==============================] - 0s 11ms/step - loss: 0.1779 - accuracy: 0.9500 - val_loss: 0.0969 - val_accuracy: 0.9667
Epoch 8/50
4/4 [==============================] - 0s 11ms/step - loss: 0.1577 - accuracy: 0.9250 - val_loss: 0.0676 - val_accuracy: 0.9667
Epoch 9/50
4/4 [==============================] - 0s 10ms/step - loss: 0.1434 - accuracy: 0.9333 - val_loss: 0.0832 - val_accuracy: 0.9667
Epoch 10/50
4/4 [==============================] - 0s 10ms/step - loss: 0.1015 - accuracy: 0.9500 - val_loss: 0.0410 - val_accuracy: 0.9667
Epoch 11/50
4/4 [==============================] - 0s 11ms/step - loss: 0.1681 - accuracy: 0.9167 - val_loss: 0.0889 - val_accuracy: 0.9667
Epoch 12/50
4/4 [==============================] - 0s 10ms/step - loss: 0.0814 - accuracy: 0.9750 - val_loss: 0.0694 - val_accuracy: 0.9667
Epoch 13/50
4/4 [==================

---
#### Question 1

Run the code above. Do you see evidence of underfitting? Overfitting? Justify your answers. ***(4 MARKS)***

**Answer: No, the consistent increase in accuracy and decrease in loss for both training and validation accuracy indicate that the network is fitting properly.  (Note to TA: If students ask why the accuracy is so low, it's because we run for only 50 epochs to save time. **

_(For TA) Marks awarded: ____ / 4_

---

#### Question 2a

Consult the documentation for the SGD optimizer [here](https://keras.io/api/optimizers/sgd/). What does the lr parameter do? ***(1 MARK)***

**Answer: The lr parameter controls how fast the network learns **

#### Question 2b

The documentation states that the momentum parameter "accelerates gradient descent in the relevant direction and dampens oscillations". Using Google or other means, illustrate what this means. ***(2 MARKS)***

**Answer: The momentum term speeds up learning while preventing the network from swinging wildly in terms of accuracy / loss, which can happen when learning rate is high **

_(For TA) Marks awarded: ____ / 3_

----

#### Question 3a

We will now play with the lr parameter. Adjust the lr parameter to the following values and record the final training and validation accuracies in the respective columns. Also observe the sequence of accuracies over the training period, and place your observation in the "remarks" column, e.g. "Progresses steadily", "some oscillation" etc. ***(3 MARKS)***

**Answer: Fill the table below **

|  lr    | Training Acc. | Validation Acc. |      Remarks      |
|:------:|---------------|-----------------|-------------------|
|0.01    | 0.6917        |       0.5667    |                   |
|0.1     | 0.950         |       0.9667    |                   |
|1.0     | 0.9667        |       0.9667    |                   |
|10.0    |               |                 |                   |
|100     |               |                 |                   |
|1000    |               |                 |                   |
|10000   |0.3667         |       0.2000    | Huge losses       |
| 100000 |               |                 |                   |


#### Question 3b

Based on your observations above, comment on the effect of small and very large learning rates on the learning. ***(2 MARKS)***

**Answer: Small learning rates result in slow learning, very large learning rates result in unstable learning that can cause huge losses and low accuracy **

_(For TA) Marks awarded: ____ / 5_

### 2.5 Using Momentum

We will now experiment with the momentum term. To do this:

    1. Change the learning rate to 0.1.
    2. Set the momentum to 0.1. Note: Do not use the Nesterov parameter - Leave it as False.
    
Run your neural network.

---

#### Question 4a

Keeping the learning rate at 0.1, complete the table below using the momentum values shown. Again record any observations in the "Remarks" column. ***(3 MARKS)***

**Answer: Fill the table below**

| momentum | Training Acc. | Validation Acc. |      Remarks      |
|:--------:|---------------|-----------------|-------------------|
|0.001     |   0.9500      |  0.93333        |                   |
|0.01      |   0.9417      |  0.9333         |                   |
|0.1       |   0.9583      |  0.9667         |                   |
|1.0       |   0.9417      |  1.0000         |                   |

#### Question 4b

Based on your observations above, does the momentum term help in learning? ***(2 MARKS)***

**Answer: The momentum appears to help with increasing learning rate especially for validation data **

_(For TA) Marks awarded: ____ / 5_

---

### 2.6 Using Raw Unscaled Data

We begin by using unscaled X and Y data. The code below will create 120 training samples and 30 testing samples (20% of the total of 150 samples):

In [23]:
X_unscaled = iris.data
Y_raw = to_categorical(y = iris.target, num_classes = 3)
X_utrain, X_utest, Y_utrain, Y_utest = train_test_split(X_unscaled, Y_raw,
                                                        test_size = 0.2,
                                                        random_state = 1)

---

#### Question 5

Create a new neural network called "nn2" below using a single hidden layer of 100 neurons. Train using the data in X_utrain, X_utest and validate with Y_utrain and Y_utest. Again use the SGD optimizer with a learning rate of 0.1 and no momentum, and train for 50 epochs. ***(3 marks)***

In [24]:
"""
Enter your code for Question 5 below
"""

# Create the neural network
nn2 = Sequential()
nn2.add(Dense(100, input_shape = (4, ), activation = 'relu'))
nn2.add(Dense(3, activation = 'softmax'))

# Create our optimizer
sgd = SGD(learning_rate = 0.1)

# 'Compile' the network to associate it with a loss function,
# an optimizer, and what metrics we want to track
nn2.compile(loss='categorical_crossentropy', optimizer=sgd, 
          metrics = 'accuracy')

nn2.fit(X_utrain, Y_utrain, shuffle = True, epochs = 50, 
      validation_data = (X_utest, Y_utest))

Epoch 1/50
4/4 [==============================] - 0s 35ms/step - loss: 2.1359 - accuracy: 0.2667 - val_loss: 1.6515 - val_accuracy: 0.2000
Epoch 2/50
1/4 [======>.......................] - ETA: 0s - loss: 1.3369 - accuracy: 0.3438

2022-05-08 16:19:54.462508: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:113] Plugin optimizer for device_type GPU is enabled.
2022-05-08 16:19:54.626532: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:113] Plugin optimizer for device_type GPU is enabled.


4/4 [==============================] - 0s 10ms/step - loss: 0.9867 - accuracy: 0.5417 - val_loss: 0.6909 - val_accuracy: 0.5667
Epoch 3/50
4/4 [==============================] - 0s 11ms/step - loss: 0.5747 - accuracy: 0.7500 - val_loss: 0.5716 - val_accuracy: 0.6000
Epoch 4/50
4/4 [==============================] - 0s 11ms/step - loss: 0.5382 - accuracy: 0.6500 - val_loss: 0.4798 - val_accuracy: 0.8000
Epoch 5/50
4/4 [==============================] - 0s 10ms/step - loss: 0.5581 - accuracy: 0.6417 - val_loss: 0.5843 - val_accuracy: 0.5667
Epoch 6/50
4/4 [==============================] - 0s 10ms/step - loss: 0.4334 - accuracy: 0.8500 - val_loss: 0.4246 - val_accuracy: 0.9667
Epoch 7/50
4/4 [==============================] - 0s 10ms/step - loss: 0.4100 - accuracy: 0.8583 - val_loss: 0.3940 - val_accuracy: 0.9333
Epoch 8/50
4/4 [==============================] - 0s 10ms/step - loss: 0.3918 - accuracy: 0.8500 - val_loss: 0.7653 - val_accuracy: 0.5667
Epoch 9/50
4/4 [======================

**(Question 5 continues)**

Observe the training and validation error. Does not scaling the input affect the training? Why do you think this is so? What is the advantage of scaling? ***(5 MARKS)***

**Answer: Not scaling the input produces similar recognition results. However scaling can prevent overflow.**

_(For TA) Marks awarded: ____ / 8_

---

## 3 Conclusion

In this lab we saw how to create a simple Dense neural network to complete the relatively simple task of learning how to classify irises according to their sepal and petal characteristics. 


---

***FOR TA ONLY***

| Question |  Marks  |
|:--------:|:-------:|
|1         |     /4  |
|2         |     /3  |
|3         |     /5  |
|4         |     /5  |
|5         |     /8  |
|Total:    |     /25 |

